src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,444 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
     }
 }
 
+pub fn enforce_comparison<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    a: &AllocatedNum<F>,
+    b: &AllocatedNum<F>,
+    diff: &AllocatedNum<F>,
+    op2: &AllocatedNum<F>,
+) -> Result<(Boolean, AllocatedPtr<F>, Boolean), SynthesisError> {
+    let double_a = add(&mut cs.namespace(|| "double a"), a, a)?;
+    // Ideally we would compute the bit decomposition for a, not for 2a,
+    // since it would be possible to use it for future purposes.
+    let double_a_bits = double_a
+        .to_bits_le_strict(&mut cs.namespace(|| "double a lsb"))
+        .unwrap();
+    let lsb_2a = double_a_bits.get(0);
+
+    let double_b = add(&mut cs.namespace(|| "double b"), b, b)?;
+    let double_b_bits = double_b
+        .to_bits_le_strict(&mut cs.namespace(|| "double b lsb"))
+        .unwrap();
+    let lsb_2b = double_b_bits.get(0);
+
+    let diff_is_zero = alloc_is_zero(&mut cs.namespace(|| "diff is zero"), diff)?;
+    let double_diff = add(&mut cs.namespace(|| "double diff"), diff, diff)?;
+    let double_diff_bits = double_diff.to_bits_le_strict(&mut cs).unwrap();
+    let lsb_2diff = double_diff_bits.get(0);
+
+    // We have that a number is defined to be negative if the parity bit (the
+    // least significant bit) is odd after doubling, meaning that the field element
+    // (after doubling) is larger than the underlying prime p that defines the
+    // field, then a modular reduction must have been carried out, changing the parity that
+    // should be even (since we multiplied by 2) to odd. In other words, we define
+    // negative numbers to be those field elements that are larger than p/2.
+    let a_is_negative = lsb_2a.unwrap();
+    let b_is_negative = lsb_2b.unwrap();
+    let diff_is_negative = lsb_2diff.unwrap();
```

Comment:
This would all be much clearer if we create an `is_negative` function/gadget, rather than manually repeating all the logic of how we test for 'negative' values inline for each check.

Reply:
done